test/hotspot/jtreg/runtime/Monitor/TestRecursiveLocking.java

```diff
@@ -0,0 +1,529 @@
+                break;
+
+            case 2:
+                syncObj.runAlternateABTest();
+                break;
+
+            default:
+                throw new RuntimeException("bad mode parameter: " + mode);
+            }
+        }
+
+        syncThread.setDone();
+        try {
+            syncThread.join();
+        } catch (InterruptedException ie) {
+            // This should not happen.
+            ie.printStackTrace();
+        }
+
+        System.out.println("INFO: main executed " + loopCount + " loops in "
+                           + n_secs + " seconds.");
+    }
+}
+
+class SyncThread extends Thread {
+    static final boolean verbose = false;  // set to true for debugging
+    private boolean done = false;
+    private boolean haveWork = false;
+    private Object obj;
+    private Object waiter = new Object();
+
+    public void run() {
+        if (verbose) System.out.println("SyncThread: running.");
+        synchronized(waiter) {
+            // Let main know that we are running:
+            if (verbose) System.out.println("SyncThread: notify main running.");
+            waiter.notify();
+
+            while (!done) {
+                if (verbose) System.out.println("SyncThread: waiting.");
+                try {
+                    waiter.wait();
+                } catch (InterruptedException ie) {
+                }
+                if (haveWork) {
+                    if (verbose) System.out.println("SyncThread: working.");
+                    synchronized(obj) {
+                    }
+                    if (verbose) System.out.println("SyncThread: worked.");
+                    haveWork = false;
+                    waiter.notify();
+                    if (verbose) System.out.println("SyncThread: notified.");
+                }
+                else if (verbose) {
+                    System.out.println("SyncThread: notified without work.");
+                }
+            }
+        }
+        if (verbose) System.out.println("SyncThread: exiting.");
+    }
+
+    public void setDone() {
+        synchronized(waiter) {
+            if (verbose) System.out.println("main: set done.");
+            done = true;
+            waiter.notify();
+        }
+    }
+
+    public void verifyCanBeSynced(Object obj) {
+        synchronized(waiter) {
+            if (verbose) System.out.println("main: queueing up work.");
+            this.obj = obj;
+            haveWork = true;
+            if (verbose) System.out.println("main: notifying SyncThread.");
+            waiter.notify();
+            if (verbose) System.out.println("main: waiting for SyncThread.");
+            while (haveWork) {
+                try {
+                    waiter.wait();
+                } catch (InterruptedException ie) {
+                }
+            }
+            if (verbose) System.out.println("main: waited for SyncThread.");
+        }
+    }
+
+    public void waitForStart() {
+        synchronized(waiter) {
```

Comment:
```suggestion
        synchronized (waiter) {
```

Reply:
Thanks for catching this. I fixed a total of 6 instances.